runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -251,6 +251,48 @@ J9::CodeGenerator::fastpathAcmpHelper(TR::Node *node, TR::TreeTop *tt, const boo
    // next treetop and then at the current one
    TR::Block* prevBlock = tt->getEnclosingBlock();
    TR::Block* targetBlock = prevBlock->splitPostGRA(tt->getNextTreeTop(), cfg, true, NULL);
+
+   // Move call tree to end of block, along with any stores of the result of the call, if they're
+   // not already at the end.  The call has to be split into its own block, but splitPostGRA above
+   // might have left register stores and stores into temporaries that must stay with the previous block
+   TR::TreeTop* prevBlockExit = prevBlock->getExit();
+   TR::TreeTop* afterTT = tt->getNextTreeTop();
+
+   if (afterTT != prevBlockExit)
+      {
+      if (trace)
+         {
+         traceMsg(comp, "Moving treetop containing node n%dn [%p] associated with acmp helper call to end of prevBlock in preparation of final block split\n", tt->getNode()->getGlobalIndex(), tt->getNode());
+         }
+
+      // Remove TreeTop for call node
+      tt->unlink(false);
+      TR::TreeTop* callTreesTail = tt;
+
+      while (afterTT != prevBlockExit)
+         {
+         TR::TreeTop* nextTT = afterTT->getNextTreeTop();
+         TR::ILOpCodes op = afterTT->getNode()->getOpCodeValue();
+
+         if ((op == TR::iRegStore || op == TR::istore) && afterTT->getNode()->getFirstChild() == node)
```

Comment:
Why we move the acmp call tree top just after node whose child is node? I think this routine can be simpler, what you need to do is move the `acmp` call helper at the end of the `prevBlock` . You might just need to get the last real tree-top and add it after it as you have just split the blocks I do not think we would have more than one exit edge from the `prevBlock`.

Reply:
@r30shah Rahil, I'm not sure I'm understanding your comment, so let me show an example of what's happening.  Here are the trees after the call to `splitPostGRA`:

```
n142n     treetop                                                                             
n141n       icall  jitAcmpHelper[#99  helper Method] [flags 0x400 0x0 ]                       
n139n         ==>aload
n140n         ==>aload
n340n     bRegStore ecx                                                                       
n14n        ==>bloadi
n343n     bRegStore edx                                                                       
n31n        ==>bloadi
n346n     bRegStore esi                                                                       
n48n        ==>bloadi
n349n     bRegStore eax                                                                       
n65n        ==>bloadi
n352n     bRegStore r14d                                                                      
n82n        ==>bloadi
n355n     istore  <temp slot 24>[#371  Auto] [flags 0x3 0x0 ]                                 
n354n       b2i                                                                               
n99n          ==>bloadi
n359n     istore  <temp slot 25>[#372  Auto] [flags 0x3 0x0 ]                                 
n358n       b2i                                                                               
n116n         ==>bloadi
n363n     istore  <temp slot 26>[#373  Auto] [flags 0x3 0x0 ]                                 
n362n       b2i                                                                               
n133n         ==>bloadi
n366n     istore  <temp slot 27>[#374  Auto] [flags 0x3 0x0 ]                                 
n141n       ==>icall
n338n     BBEnd </block_2> ===== ()
```

I'm trying to move the `TR::TreeTop`s containing nodes n142n and n366n.  I hadn't wanted to make any assumptions about how many stores there might be for the value of the `icall` or whether they would appear after all of the other stores.

Reply:
Thanks a lot for the example. So when you split the block, is `n141n` referenced after split point?

Reply:
Okk, Sorry, now it is clear to me.  So far it seems good to me, let me give it another look.